File: repository/src/main/java/com/rxdroid/repository/model/Resource.java
package com.rxdroid.repository.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.rxdroid.api.error.RequestError;

public class Resource<T> {

    @NonNull
    public final Status status;
    @Nullable
    public final T data;
    @Nullable
    public final RequestError requestError;

    public Resource(@NonNull Status status, @Nullable T data, @Nullable RequestError requestError) {
        this.status = status;
        this.data = data;
        this.requestError = requestError;
    }

    public static <T> Resource<T> success(@NonNull T data) {
        return new Resource<>(Status.SUCCESS, data, null);
    }

    public static <T> Resource<T> error(RequestError requestError, @Nullable T data) {
        return new Resource<>(Status.ERROR, data, requestError);
    }

    public static <T> Resource<T> loading(@Nullable T data) {
        return new Resource<>(Status.LOADING, data, null);
    }
}

File: repository/src/main/java/com/rxdroid/repository/model/Resource.kt
package com.rxdroid.repository.model

import com.rxdroid.api.error.RequestError

class Resource<out T>(val status: Status, val data: T?, val requestError: RequestError?) {

    companion object {

        fun <T> success(data: T): Resource<T> {
            return Resource(Status.SUCCESS, data, null)
        }

        fun <T> error(requestError: RequestError): Resource<T> {
            return Resource(Status.ERROR, null, requestError)
        }

        fun <T> loading(): Resource<T> {
            return Resource(Status.LOADING, null, null)
        }
    }
}
